server/src/main/java/org/elasticsearch/index/mapper/MapperServiceUpgraders.java

```diff
@@ -0,0 +1,79 @@
+import org.elasticsearch.common.compress.CompressedXContent;
+import org.elasticsearch.common.xcontent.XContentHelper;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * Contains a collection of helper methods for index mapping upgrades
+ */
+public class MapperServiceUpgraders {
+    @SuppressWarnings("unchecked")
+    // package private for testing
+    static Map<String, Object> getMappingProperties(Map<String, Object> mappingMap, String type) {
+        // Check for MappingMap without type and ignore it
+        if (mappingMap.size() != 1) {
+            return null;
+        }
+        // Fetch the mappings for the encoded type
+        Map<String, Object> mappingType = (Map<String, Object>) mappingMap.get(type);
+        if (mappingType == null) {
+            return null;
+        }
+
+        return (Map<String, Object>) mappingType.get("properties");
+    }
+
+    /**
+     * Upgrades the format field of mapping properties
+     * <p>
+     * This method checks is the new mapping source has a different format than the original mapping, and if it
+     * does, then it returns the original mapping source with updated format field. We cannot simply replace the
+     * MappingMetadata mappingSource with the new parsedSource, because the parsed source contains additional
+     * properties which cannot be serialized and reread.
+     *
+     * @param mappingMetadata the current mapping metadata
+     * @param parsedSource the mapping source as it was parsed
+     */
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    public static CompressedXContent upgradeFormatsIfNeeded(MappingMetadata mappingMetadata, CompressedXContent parsedSource) {
+        Map<String, Object> sourceMap = mappingMetadata.rawSourceAsMap();
+        Map<String, Object> newMap = XContentHelper.convertToMap(parsedSource.compressedReference(), true).v2();
+
+        Map<String, Object> properties = getMappingProperties(sourceMap, mappingMetadata.type());
+        Map<String, Object> parsedProperties = getMappingProperties(newMap, mappingMetadata.type());
+        if (properties != null && parsedProperties != null) {
+            for (var property : properties.entrySet()) {
+                if (property.getValue()instanceof Map map) {
+                    Object format = map.get("format");
+                    if (format != null) {
+                        var newProperty = parsedProperties.get(property.getKey());
+                        if (newProperty instanceof Map parsedMap) {
+                            Object parsedFormat = parsedMap.get("format");
+                            if (parsedFormat != null && parsedFormat.equals(format) == false) {
+                                map.put("format", parsedFormat);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        try {
+            return new CompressedXContent(sourceMap);
```

Comment:
Could you keep track of whether any change was made above, so that the original parsedSource can be returned here and reference equality could be used by the caller to detect changes? That would make it less expensive when there aren't any changes (which is most of the time).